fix: Unable to build workspaces with circular dependencies

oliversalzburg opened the pull request:
A simple `Map` to keep track of already planned workspaces. I'm not too sure with the approach, but it seems to work as expected.

It actually works better than expected, because now a lot more projects seem to be built at the same time. Which makes me think something is actually not right. Or this is actually the correct result when having circular dependencies, as the order isn't clear.

This also includes some random cleanups and fixes already present in other PRs. I can rebase later if needed.

Fixes #64

oliversalzburg commented:
After some testing, I can confidently say, this isn't the fix :D

Resulted in required depedency workspaces not being built in CI :( I'll try again, but I might have to research how to properly run/debug the plugin.

oliversalzburg commented:
I've rewritten the recursive change-detection in another codebase. First results are promising. Changes are detected properly and dependency loops are detected.

Once other PRs settle, I'd like to try and integrate this.

ojkelly commented:
Thanks for digging into this. 

The general approach for building (if I remember correctly) is to build a tree from the current target (or all targets if at root). Then starting from the first targets build as many first level dependent targets as we have threads.

circular deps do absolutely throw a spanner in this, because it's mathematically impossible to solve. I'm not exactly sure what the solution is, however I'm keen to see if you can get the build plan to a state that works for you're circular deps.

With the circular deps, are they dependent on each others output, or do they just reference each others source/assets?
if it's the former, I think it's impossible to solve. But the latter is pretty common in the JS world, so it would be good to have a good solution for that scenario.

oliversalzburg commented:
That's actually very reassuring to me, thanks 😀 My case is the latter case. I have a bunch of microservices that contain a proxy class, which allows type-safe communication with the service. This is not a build artifact, but generated manually on-demand. So the other services just want that proxy class and don't care about any build output.

I totally agree that a dependency on build output is probably impossible to solve or would at least lead down a very bad rabbithole.

I was able to solve this for my own use case, but I'm not sure how well it'll work in the long run (still testing). I've approached this in multiple stages and built the approach as a standalone instead of a yarn plugin for easier debugging.

1. Enumerate all targets. This is either all (non-top-level) workspaces (default) or a set of workspaces (targetted).
2. For each target, evaluate _freshness_. A target is fresh if:
	1. All files in the target's sources are not changed since last audit.
	2. All files in dependencies of the target are not change since last audit.
3. The information is collected into a report. The collector detects if a dependency branch loops back to a parent and cuts off execution at that point.
4. The resulting report will clearly indicate which target is fresh and which target requires to be rebuilt, including the cause. A report inspector can then be used to unroll the report into build instructions for all non-fresh elements in the project.

![image](https://user-images.githubusercontent.com/1658949/104006514-e1399380-51a6-11eb-8efc-3245cd1799cf.png)

I will try to put this on GH in a shape that we can both properly inspect and discuss it. So that you can get a better idea.

oliversalzburg commented:
I pushed the code to https://github.com/oliversalzburg/yarn-build-auditor We can also discuss it on Discord, if you like and have the time 😊

ojkelly commented:
I've just had a look through the repo -- looks good. I think that would almost solve #8, which is an issue I added early on, because I thought this part of the plugin still needed work.

I'm happy to get this merged in when you're ready.

oliversalzburg commented:
Cool. I'll try to move it into the yarn.build codebase and wire everything up. I'll also see if I can make the `yarn query` command work to some degree. It might take me a couple of days though 😊

ojkelly commented:
no worries. probably `yarn build --query` just because I'm weary about stealing too many commands from users, who may have other ideas for them.

oliversalzburg commented:
Just FYI, I have since resolved the circular dependency in my own code base due to other issues. I've started work on integrating the new auditor anyway, but since things are working relatively well for now, I've focused on other issues.

Before I abandon the idea, I'll definitely close the issue though 😅

arcanis commented:
Just hit this problem as well 😄

In Yarn itself, we partially solve this problem with two caveats:

- We have `--topological`, which only accounts for regular dependencies. This means that if A depends on B via a dev dep, it won't need to wait for B to build before being built itself. This alone covers most grounds, since circular dependencies are often runtime-only and don't have actual impact on the build.

- We remove from the traversal packages that have no build script - since even if they have a circular dependency, in practice there isn't a build-time cycle there.

oliversalzburg commented:
I've looked into this once more and I don't think I'll be able to integrate my approach without basically rewriting the entire plugin (which I don't plan on doing). Sorry.
